lib/benchee/benchmark/runner.ex

```diff
@@ -120,35 +177,56 @@ defmodule Benchee.Benchmark.Runner do
+    run_time = iteration_time(scenario, scenario_context)
     updated_scenario = %Scenario{scenario | run_times: [run_time | run_times]}
     updated_context =
       %ScenarioContext{scenario_context | current_time: current_time()}
     do_benchmark(updated_scenario, updated_context)
   end
 
-  defp measure_call(scenario, scenario_context = %ScenarioContext{
-                      num_iterations: num_iterations
-                    }) do
-    measure_call_n_times(scenario, scenario_context) / num_iterations
+  defp iteration_time(scenario, scenario_context = %ScenarioContext{
+                                  num_iterations: num_iterations
+                                }) do
+    measure_iteration(scenario, scenario_context) / num_iterations
   end
 
-  @no_input Benchmark.no_input()
-  defp measure_call_n_times(%Scenario{function: function, input: @no_input},
-                            %ScenarioContext{num_iterations: num_iterations}) do
-    {microseconds, _return_value} = :timer.tc fn ->
-      RepeatN.repeat_n(function, num_iterations)
-    end
-
+  defp measure_iteration(scenario, scenario_context = %ScenarioContext{
+                                     num_iterations: 1,
+                                     benchmarking_function: function
+                                   }) do
+    run_before_each(scenario, scenario_context)
+    {microseconds, _return_value} = :timer.tc function
+    run_after_each(scenario, scenario_context)
     microseconds
   end
-  defp measure_call_n_times(%Scenario{function: function, input: input},
-                            %ScenarioContext{num_iterations: num_iterations}) do
-    fun_with_input = fn -> function.(input) end
-    {microseconds, _return_value} = :timer.tc fn ->
-      RepeatN.repeat_n(fun_with_input, num_iterations)
-    end
-
+  defp measure_iteration(_scenario, %ScenarioContext{
+                          num_iterations: iterations,
+                          benchmarking_function: function
+                        }) when iterations > 1 do
+    # When we have more than one iteration, then the repetition and calling
+    # of hooks is already included in the function, for reference/reasoning see
+    # `build_benchmarking_function/2`
+    {microseconds, _return_value} = :timer.tc function
     microseconds
   end
+
+  defp run_before_each(%{
+                         before_each: local_before_each
+                       },
+                       %{
+                         config: %{before_each: global_before_each}
+                       }) do
+    if global_before_each, do: global_before_each.()
+    if local_before_each,  do: local_before_each.()
+  end
+
+  defp run_after_each(%{
+                        after_each: local_after_each
+                      },
+                      %{
+                        config: %{after_each: global_after_each}
+                      }) do
+    if global_after_each, do: global_after_each.()
```

Comment:
I think the way RSpec does it (which is what I'm taking as inspiration here), the "outer" scope comes second. See here: https://relishapp.com/rspec/rspec-core/v/2-0/docs/hooks/before-and-after-hooks#before/after-all-blocks-are-run-once

So, would we consider the global function to be the "outer", and run that after the local `after_each` here?

Reply:
definitely it's actually the way I wanted to do that and the way I layed it out in the tests but my test helper doesn't verify the order of meessages, sadly... maybe I'll patch that in.

Good catch either way :+1: